File: algorithm/python/BOJ15651.py
# from https://www.acmicpc.net/problem/15651

import sys

N, M = map(int, sys.stdin.readline().split())

l = list(map(lambda x: x, range(1, N + 1)))
result = [0] * M


def permutation(cnt):
    if cnt == M:
        for num in result:
            print(num, end=" ")
        print()

        return

    for i in range(0, len(l)):
        result[cnt] = l[i]
        permutation(cnt + 1)


permutation(0)

File: algorithm/typescript/BOJ15651.ts
// from https://www.acmicpc.net/problem/15651

import fs = require("fs");

const input = fs.readFileSync("/dev/stdin").toString().split(" ");
const N: number = parseInt(input[0]);
const M: number = parseInt(input[1]);

const inputList = [];
const outputList = [];
const result = [];

for (let i = 0; i < N; i++) {
  inputList.push(i + 1);
}

function permutation(cnt: number): void { 
  if (cnt == M) {
    for (let i = 0; i < M; i++) {
      result.push(`${outputList[i]} `);
    }
    result.push(`\n`);
    return;
  }

  for (let i = 0; i < N; i++) {
    outputList[cnt] = inputList[i];
    permutation(cnt + 1);
  }
}

permutation(0);
console.log(result.join(""));